File: python/2015/07/solution.py
import common
import operator
import re

lines = common.read_file('2015/07/data.txt').splitlines()

values = dict()
gates = []

def bit_not(val):
    return val ^ 0b1111111111111111

def passthrough(val):
    return val

def get_val(val_spec):
    if re.match('\\d+', val_spec):
        return int(val_spec)
    if val_spec in values.keys():
        return values[val_spec]
    return None

def set_val(name, val):
    values[name] = val
    if name == 'a':
        print(val)

seen = set()

def propagate_val(name):
    fulfilled = []
    for g in gates:
        if name in g[0] and g[2] not in values.keys():
            vals = list(map(get_val, g[0]))
            vals_not_none = list(map(lambda x: x is not None, vals))
            if not all(vals_not_none):
                continue
            result = g[1](*vals)
            set_val(g[2], result)
            fulfilled.append(g)

    for g in fulfilled:
        propagate_val(g[2])

def op_for_name(name):
    if name == 'AND':
        return operator.and_
    elif name == 'OR':
        return operator.or_
    elif name == 'LSHIFT':
        return operator.lshift
    else: # RSHIFT
        return operator.rshift

for line in lines:
    segments = line.split()
    name = segments[-1]
    if len(segments) == 3: # just a value provided
        gates.append((( segments[0], ), passthrough, name))
        val = get_val(segments[0])
        if val is not None:
            set_val(name, val)
            propagate_val(name)
    elif len(segments) == 4: # NOT
        gates.append((( segments[1], ), bit_not, name))
        val = get_val(segments[1])
        if val is not None:
            set_val(name, bit_not(val))
            propagate_val(name)
    else: # two-arg operators
        op = op_for_name(segments[1])
        gates.append((( segments[0], segments[2] ), op, name))
        val1 = get_val(segments[0])
        val2 = get_val(segments[2])
        if val1 is not None and val2 is not None:
            set_val(name, op(val1, val2))
            propagate_val(name)

# part 2
values = dict(b=values['a'])

for g in gates:
    if g[2] not in values:
        vals = list(map(get_val, g[0]))
        vals_not_none = list(map(lambda x: x is not None, vals))
        if not all(vals_not_none):
            continue
        result = g[1](*vals)
        set_val(g[2], result)
        propagate_val(g[2])

File: go/2015/07/solution.go
package main

import (
	"aoc/utils"
	"fmt"
	"regexp"
	"slices"
	"strings"
)

type (
	Op     func([]*int) int
	Values map[string]int
	Gate   struct {
		inputs []string
		op     Op
		output string
	}
)

func bitNot(val []*int) int {
	return *val[0] ^ 0b1111111111111111
}

func passthrough(val []*int) int {
	return *val[0]
}

func getVal(v Values, valSpec string) *int {
	if matched, _ := regexp.MatchString("\\d+", valSpec); matched {
		r := new(int)
		*r = utils.Atoi(valSpec)
		return r
	} else if _, ok := v[valSpec]; ok {
		r := new(int)
		*r = v[valSpec]
		return r
	}
	return nil
}

func setVal(values Values, name string, val int) {
	values[name] = val
	if name == "a" {
		fmt.Println(val)
	}
}

func propagateVal(v Values, gates []Gate, name string) {
	fulfilled := []Gate{}
	for _, g := range gates {
		if slices.Contains(g.inputs, name) {
			if _, ok := v[g.output]; !ok {
				vals := utils.Map(g.inputs, func(s string) *int { return getVal(v, s) })
				valsNotNone := utils.Map(vals, func(i *int) bool { return i != nil })
				if !utils.All(valsNotNone) {
					continue
				}
				result := g.op(vals)
				setVal(v, g.output, result)
				fulfilled = append(fulfilled, g)
			}
		}
	}

	for _, g := range fulfilled {
		propagateVal(v, gates, g.output)
	}
}

func opForName(name string) Op {
	if name == "AND" {
		return func(v []*int) int { return *v[0] & *v[1] }
	} else if name == "OR" {
		return func(v []*int) int { return *v[0] | *v[1] }
	} else if name == "LSHIFT" {
		return func(v []*int) int { return *v[0] << *v[1] }
	} else { // RSHIFT
		return func(v []*int) int { return *v[0] >> *v[1] }
	}
}

func main() {
	lines := utils.ReadLines()

	values := make(Values)
	gates := []Gate{}

	for _, line := range lines {
		segments := strings.Split(line, " ")
		name := segments[len(segments)-1]
		if len(segments) == 3 { // just a value provided
			gates = append(gates, Gate{[]string{segments[0]}, passthrough, name})
			val := getVal(values, segments[0])
			if val != nil {
				setVal(values, name, *val)
				propagateVal(values, gates, name)
			}
		} else if len(segments) == 4 { // NOT
			gates = append(gates, Gate{[]string{segments[1]}, bitNot, name})
			val := getVal(values, segments[1])
			if val != nil {
				setVal(values, name, bitNot([]*int{val}))
				propagateVal(values, gates, name)
			}
		} else { // two-arg operators
			op := opForName(segments[1])
			gates = append(gates, Gate{[]string{segments[0], segments[2]}, op, name})
			val1 := getVal(values, segments[0])
			val2 := getVal(values, segments[2])
			if val1 != nil && val2 != nil {
				setVal(values, name, op([]*int{val1, val2}))
				propagateVal(values, gates, name)
			}
		}
	}

	// part 2
	aVal := values["a"]
	values = make(Values)
	values["b"] = aVal

	for _, g := range gates {
		if _, ok := values[g.output]; !ok {
			vals := utils.Map(g.inputs, func(s string) *int { return getVal(values, s) })
			valsNotNone := utils.Map(vals, func(i *int) bool { return i != nil })
			if !utils.All(valsNotNone) {
				continue
			}
			result := g.op(vals)
			setVal(values, g.output, result)
			propagateVal(values, gates, g.output)
		}
	}
}
